compiler/bindings/python/iree/compiler/tools/onnx.py

```diff
@@ -0,0 +1,165 @@
+    try:
+        onnx.checker.check_model(raw_model)
+        inferred_model = onnx.shape_inference.infer_shapes(
+            raw_model, data_prop=data_prop
+        )
+        return inferred_model
+    except ValueError:
+        pass
+
+    # Model is too big for in-memory inference: do file-based shape inference
+    # to a temp file.
+    # Make a temp dir for all the temp files we'll be generating as a side
+    # effect of infering shapes. For now, the only file is a new .onnx holding
+    # the revised model with shapes.
+    with tempfile.TemporaryDirectory() as temp_dir_name:
+        temp_dir_path = Path(temp_dir_name)
+        temp_input_file = temp_dir_path + "/" + "temp.onnx"
+        temp_inferred_file = temp_dir_path + "/" + "temp-inferred.onnx"
+        onnx.save(raw_model, temp_input_file)
+        onnx.shape_inference.infer_shapes_path(
+            temp_input_file, temp_inferred_file, data_prop=data_prop
+        )
+
+        # Load the temp file and the external data.
+        inferred_model = onnx.load(
+            temp_inferred_file, load_external_data=False)
+        data_dir = Path(temp_input_file if data_dir is None else data_dir)
+        onnx.load_external_data_for_model(inferred_model, str(data_dir))
+
+        return inferred_model
+
+
+@dataclass
+class ImportOptions(CompilerOptions):
+    input_type: Union[InputType, str] = InputType.ONNX
+    entry_point_name: Optional[str] = None
+    module_name: Optional[str] = None
+    import_only: bool = False
+    save_temp_iree_input: Optional[str] = None
+    verify_module: bool = False
+    use_bytecode: bool = False
+
+
+def compile_saved_model(model_path: IO[bytes] | str | os.PathLike, **kwargs):
+    options = ImportOptions(**kwargs)
+
+    with TempFileSaver.implicit() as tfs, tempfile.TemporaryDirectory() as tmpdir:
+        if options.import_only and options.output_file:
+            # Importing to a file and stopping, write to that file directly.
+            onnx_iree_input = options.output_file
+        elif options.save_temp_iree_input:
+            # Saving the file, use tfs.
+            onnx_iree_input = tfs.alloc_optional(
+                "onnx-iree-input.mlirbc", export_as=options.save_temp_iree_input
+            )
+        else:
+            # Not saving the file, so generate a loose temp file without tfs.
+            onnx_iree_input = os.path.join(tmpdir, "onnx-iree-input.mlirbc")
+            # onnx_iree_input = io.BytesIO()
+
+        _IR_MIN_VERSION = 17
+
+        # convert onnx model to version if needed 17
+        original_model = onnx.load_model(model_path)
+        if original_model.ir_version < _IR_MIN_VERSION:
+            logger.info("Converting onnx model opset version from %s to %s",
+                        original_model.ir_version, _IR_MIN_VERSION)
+            converted_model = onnx.version_converter.convert_version(
+                original_model, _IR_MIN_VERSION)
```

Comment:
This can fail, so we should wrap with some error handling and log a helpful error message if possible.

I'm not sure what the best practices are, but here's some code I wrote: https://github.com/iree-org/iree-test-suites/blob/9f2ca6dde62a628adc8ae0bbdf3afe1f0a883b9b/onnx_ops/import_onnx_tests.py#L75-L93

Reply:
[ec2ef59](https://github.com/iree-org/iree/pull/18365/commits/ec2ef59f89105b86a87f5005dc82ec453af2b7df)